Add first draft material for an nf-core module writing workshop (#3140)
nf-core/website: sites/docs/src/content/docs/tutorials/nf-core_training/writing-nf-core-modules/chapter-6-testing-your-module.md

The test block declaration consists of three main sections:

- The title of the test
- The when block (defining inputs)
- The then block (defining outputs)

The boilerplate template comes with example code for a single input file as input, and tests both that the module successfully passes with no errors, and generates a 'snapshot' of md5sums of the contents of all output channels of the module.
In addition, the boilerplate template also includes a stub runs test.
All nf-core modules require a stub-run test, however you do not need to change this except for the test title, and the inputs (the 'then' block) so they match that the first test.

For writing the test, you need to follow the following steps.

First, you must update the name of the test to make it distinct.
Typically, at a minimum this will consist of the organism of the test data you will use, and the file format of the primary input file.
You can then also provide an additional keyword or short sentence of what configuration the given test will be testing, for example:

```nextflow
test("sarscov2 - fastq - pairedend")
```

or

```nextflow
test("sarscov2 - fastq - bakta annotation input")
```

> Second, within the 'when' block you then need to specify each input channel to your module using the notation of `input[0]` for the first input channel, `input[1]`, for the second, and so forth.

Reviewer: ```suggestion
Second, within the 'when' block you need to specify each input channel to your module using the notation of `input[index]`, i.e. `input[0]` for the first input channel, `input[1]`, for the second, and so forth.
```

This 'when' block can be filled with standard Nextflow code - e.g. using a `Channel` factory to create a channel, use operators, and so on, as you normally would in an Nextflow pipeline.

> Make sure that the input into each channel `input[]` variable matches the channel restructure of your module (e.g. if there is a meta map)

Reviewer: ```suggestion
Make sure that the input into each channel `input[index]` variable matches the channel restructure of your module (e.g. if there is a meta map)
```

This is where you will need to write 'assertions', i.e., telling nf-test what to compare between test runs.
By default, the example in the boilerplate template code will check the contents of every single output file of the channel.
However, very often you may find some tools produce variable output files between runs (a common one is logs which include timestamps).
You can expand the `snapshot()` function to explicitly refer to specific channels, and test each one in different ways.

```nextflow
then {
    assertAll(
        { assert process.success },
        { assert snapshot(
            process.out.metrics,
            process.out.versions,
            file(process.out.qc_report[0][1]).name,
        ).match()
        }
    )
}
```

For example, in the example above you can see that the `metrics` and `versions` output channels do not vary in their md5sum content.
However the `qc_report` does, therefore instead of using the (default) `md5sums` check, we change the 'assertion' so that we compare that name of the files in that channel does not change between runs.

There are many different `assert` methods.

> For a more comprehensive list of different nf-test assertions, see the dedicated [nf-core documenation](https://nf-co.re/docs/contributing/nf-test/assertions) page.

Reviewer: ```suggestion
For a more comprehensive list of different nf-test assertions, see the dedicated [nf-core documentation](https://nf-co.re/docs/contributing/nf-test/assertions) page.
```

We generally recommend the to test files with the following methods in order of preference:

1. 😃 `md5sum`s
2. 🙂 String contents within a file
3. 🫤 File name / file existence

If you have to vary the types of tests per channel, make sure to always include testing of the `process.out.versions` channel!

Finally, once you've completed both the `then` and `when` blocks, you can copy the structure of this first test to each subsequent test case you need to cover - updating the name test and contents.

As a guide, you should try and have as many tests so you test as many configurations as possible, but at a minimum at least enough tests so that you test input files for every input and output channel (mandatory and optional) at least once.

### The `setup` block (optional)

> This optional section is where you can specify module(s) to execute _before_ your new module itself

Reviewer: ```suggestion
This optional section is where you can specify module(s) to execute _before_ your new module.
```

The outputs of this upstream module can then be used as input for your new module.
Note that it is not included by default in the boilerplate template code.

```nextflow
    setup {
        run ("UNTAR") {
            script "../../../untar/main.nf"
            process {
                """
                input[0] = [[],file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/db/kraken2_bracken.tar.gz', checkIfExists: true)]
                """
            }
        }
    }
```

The location of a setup block can vary.
You can either specify this before all tests, so you can re-use the same output of this upstream module in all subsequent tests.
Alternatively you can place it before the 'when' block of each test within the test block themselves.
In this case the the setup block will only be executed when the given test is executed.

> The difference between setup block and the test block (see above) is that the output of modules in the setup block will _not_ be evaluated by the test.

Reviewer: Would consider using _asserted_ over _evaluated_ here to make clear that it will run, but is not part of the assertion

You can fill this block in just the same way as the `test` block, except you must explicitly specify the script path of the upstream module in each setup block.

> Otherwise you specify the inputs wotj the same `input[0]`, `input[1]` etc. channel syntax, and using the URLs to the nf-core test-dataset repository as before.

Reviewer: ```suggestion
Otherwise you specify the inputs with the same `input[0]`, `input[1]` etc. channel syntax, and using the URLs to the nf-core test-dataset repository as before.
```

Note that we generally discourage the use of setup blocks as they increase the runtime of tests.

> However it can be useful when a module requires inputs with large file-sizes that are too large for the nf-core/test-datasets repository, or the upstream module is extremely quick.

Reviewer: ```suggestion
However they can be useful when a module requires inputs with large file-sizes that are too large for the nf-core/test-datasets repository, or the upstream module is extremely quick.
```

                input[1] = []
                input[2] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.success },
                { assert snapshot(
                    process.out.vcf,
                    process.out.versions
                ).match() }
            )
        }

    }
```

:::

## The `nextflow.config` file (optional)

This is an optional file that is _not_ automatically generated with the boilerplate template files.
It sits alongside the `main.nf.test` and `main.nf.test.snap` file within the `tests/` directory.

The purpose of this file is to provide additional settings, e.g. specifying `ext.args` for modules when you want to expand the number of configurations to test.
For example, you could use it if you need to give an optional parameter to produce an additional optional output file.
This can also be use to set up different options for when you need to import a module multiple times (typically in `setup` blocks).

If you need this functionality, you will need to create the file yourself.

> The file should sit alongside in the same directory as `tests/main.nf.test`:

Reviewer: Redundant with L390

This file is _not_ automatically generated with the boilerplate template files.
It sits alongside the `main.nf.test` and file within the `tests/` directory, and is automatically generated for you when you run the `main.nf.test` script file with `nf-test` the first time.

```tree {6}
├── environment.yml
├── main.nf
├── meta.yml
└── tests/
    ├── main.nf.test
    ├── main.nf.test.snap
    └── nextflow.config
```

You can generate the file by running nf-core tools command:

```bash
nf-core modules test <tool>/<subcommand>
```

and after setting the options via the prompts, the tests will run twice and tell you if the output of each tests were the same or not, in addition to generating the snapshot file (in the first time)

The generated file includes the 'snapshot' output of nf-test.
This snapshot is what will be compared against for each subsequent test run (e.g., comparing the `md5sum` or presence of a string in the output between two given tests).

You never need to manually modify this file, as it will also be automatically updated by nf-test.
If you find variability or an issue in your snapshot, you can re-run with

```bash
nf-core modules test <tool>/<subcommand> --update
```

> If you need more information about the test run, you can also provide `--debug` to see the full nf-test stdout log.

Reviewer: Everything below here seems to be very specific for testing. I am a bit concerned that this might be a bit much for someone trying to understand how to write a module with no background. I guess if this is in-scope depends on what the scope of this document and if there is another separate workshop for testing.

Author: It probably should be it's own workshop, but it is required for the module so I've tried to roughly introdce it it here - I will ad a warning at the top of the page to that effect

It is important to manually inspect the snapshot file once it has been generated.
You need to make sure you have no unexpected snapshot assertion outputs, even if the tests 'pass' (e.g. empty file `md5sum`, or missing assertions etc.).

If you find issues, and need to inspect the output files of the module itself, we can check the working directory of the module during the test.

To inspect this, in the console when you're running each test, before the name of each test you should get a little 'hash' string:

```bash
│   Test [528b411a] 'candidatus_portiera_aleyrodidarum proteome [fasta]'       │
│ PASSED (7.762s)
```

With this hash string you can change into `.nf-test/tests/<hash string>/work` (make sure to autocomplete with your TAB key to get the full hash) and you can find all the standard Nextflow working directories of the module's test run.

> Within here you should change into each directory until you find the one of the module itself, and check the contents of each output file are as expected to ensure your snapshot has been generated correctly.

Reviewer: ```suggestion
Here, you should change into each directory until you find the one of the module itself, and check that the contents of each output file are as expected to ensure your snapshot has been generated correctly.
```
